sw/host/opentitanlib/src/io/emu.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transport::Result;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use thiserror::Error;
+
+/// Error related to the `Emulator` trait. These error messages will be printed in the context of
+/// a TransportError::EmuError, that is "Emulator error: {}".  So including the words "error" or
+/// "Emulator" in texts below will probably be redundant.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum EmuError {
+    #[error("Invalid argument name {0}")]
+    InvalidArgumetName(String),
+    #[error("Argument {0} has invalid value {1}")]
+    InvalidArgumentValue(String, String),
+    #[error("Start failed with cause {0}")]
+    StartFailureCause(String),
+    #[error("Stop failed with cause {0}")]
+    StopFailureCause(String),
+    #[error("Timeout during performing command: {0} timeout: {1}, elsped: {2}")]
+    Timeout(String, u64, u64),
+    #[error("Can't restore resource to initial state: {0}")]
+    ResetError(String),
+    #[error("Runtime error {0}")]
+    RuntimeError(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum EmuValue {
+    Empty,
+    String(String),
+    FilePath(String),
+    StringList(Vec<String>),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
+pub enum EmuState {
+    Off,
+    On,
+    Busy,
+    Error,
+}
+
+/// A trait which represents a `Emulator` instance
+pub trait Emulator {
+    /// Acquires state of emulator
+    fn get_state(&self) -> Result<EmuState>;
+
+    /// Waits for the specified amount of time passed in `timeout_ms`
+    /// for the Emulator state to change to the value provided in parameter `state`.
+    fn wait_for_state(&self, timeout_ms: u64, state: EmuState) -> Result<()>;
```

Comment:
I assume a common use case would be that the current state is `Busy`, and I could invoke this method to wait 15 seconds for the state to become `On`.  However, if after 3 seconds, the state changes to `Error`, I do not think there is much point in waiting further 12 seconds, as we know that the state would not likely become `On` by itself.

I would suggest that, we change the documentation such that it is clear that this method waits only while the state is `Busy`, and returns immediatly with an `Err()` if the state changes to anything other than what is specified in the parameter.  Alternatively, we could drop the `state` parameter, and make it return `Result<EmuState>`, documenting that it will never return Ok(Busy), and that it will wait for up to the timeout for the state to become anything other than `Busy`.

Reply:
```
On       |                     /|-------------------\
Busy     |          /---------/ |                    \---\
Off -----|-----(*)-/            |                         \---------
         |<---------wait------->|
        Off                     On        (*)-issuing start command
```

            
The idea is that you issue command `wait_for_state` in any point of time and Emulator current state could by any of
4 valid state. For example when you issue command `start` you could be in `Error` or `Off` state even  `Busy` (in that case `start` will wait in queue, Emulator will try to execute when it finish current command)

This instruction is like a barrier - all command executed after this barrier will be executed after transition to defined state.
I think it will be clean to understand when PR with 

My alternative is add 2 function waiting for start and stop. But current `wait_for_status` is more flexible and simpler to implement.

Reply:
Changed to `wait_until_busy(..)`.

Reply:
Hm, I assumed that immediately after calling start(), the state would be Busy, and that the typical sequence would be someone calling start(), and then immediately afterwards calling wait_until_not_busy().  I find the sequence in the above diagram odd, in that why would wait() be called before start() (and start() would have to be on a separate thread).
If we want to support wait() being called after start(), then we have to prepare for the case that the state has already transitioned to Busy and further to On before wait(), is called.  And in that case, we want it to return immediately.  For that reason, I do not like the comment that states that wait() returns after the first _transition_ that goes to something other than Busy.  Rather, I think wait should return at the first time when the state is not Busy, even if that is immediately.